File: sras-ui/src/app/services/project.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { Project, ProjectRequirement } from '../models/project.model';

@Injectable({ providedIn: 'root' })
export class ProjectService {
  private readonly apiUrl = `${environment.apiUrl}/projects`;

  constructor(private http: HttpClient) {}

  getAll(): Observable<Project[]> {
    return this.http.get<Project[]>(this.apiUrl);
  }

  getById(id: number): Observable<Project> {
    return this.http.get<Project>(`${this.apiUrl}/${id}`);
  }

  create(project: Project): Observable<Project> {
    return this.http.post<Project>(this.apiUrl, project);
  }

  update(id: number, project: Project): Observable<Project> {
    return this.http.put<Project>(`${this.apiUrl}/${id}`, project);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getRequirements(projectId: number): Observable<ProjectRequirement[]> {
    return this.http.get<ProjectRequirement[]>(`${this.apiUrl}/${projectId}/requirements`);
  }

  addRequirement(projectId: number, req: ProjectRequirement): Observable<ProjectRequirement> {
    return this.http.post<ProjectRequirement>(`${this.apiUrl}/${projectId}/requirements`, req);
  }

  updateRequirement(reqId: number, req: ProjectRequirement): Observable<ProjectRequirement> {
    return this.http.put<ProjectRequirement>(`${this.apiUrl}/requirements/${reqId}`, req);
  }

  deleteRequirement(reqId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/requirements/${reqId}`);
  }
}

File: sras-backend/src/main/java/com/pss/SRAS/services/ProjectService.java
package com.pss.SRAS.services;

import com.pss.SRAS.models.Project;
import com.pss.SRAS.models.ProjectRequirement;
import com.pss.SRAS.repositories.ProjectRepository;
import com.pss.SRAS.repositories.ProjectRequirementRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.NoSuchElementException;

@Service
@RequiredArgsConstructor
public class ProjectService {

    private final ProjectRepository projectRepository;
    private final ProjectRequirementRepository requirementRepository;

    public List<Project> getAllProjects() {
        return projectRepository.findAll();
    }

    public Project getById(Long id) {
        return projectRepository.findById(id)
                .orElseThrow(() -> new NoSuchElementException("Project not found: " + id));
    }

    @Transactional
    public Project create(Project project) {
        if (project.getProjectRequirements() != null) {
            project.getProjectRequirements().forEach(r -> r.setProject(project));
        }
        return projectRepository.save(project);
    }

    @Transactional
    public Project update(Long id, Project updated) {
        Project existing = getById(id);
        existing.setProjectName(updated.getProjectName());
        existing.setDomain(updated.getDomain());
        existing.setLocationPreferences(updated.getLocationPreferences());
        existing.setProjectManager(updated.getProjectManager());

        existing.getProjectRequirements().clear();
        if (updated.getProjectRequirements() != null) {
            updated.getProjectRequirements().forEach(r -> {
                r.setProject(existing);
                existing.getProjectRequirements().add(r);
            });
        }
        return projectRepository.save(existing);
    }

    @Transactional
    public void delete(Long id) {
        if (!projectRepository.existsById(id)) {
            throw new NoSuchElementException("Project not found: " + id);
        }
        projectRepository.deleteById(id);
    }

    // ─── Requirements ─────────────────────────────────────────────────────────

    public List<ProjectRequirement> getRequirements(Long projectId) {
        return requirementRepository.findByProjectId(projectId);
    }

    @Transactional
    public ProjectRequirement addRequirement(Long projectId, ProjectRequirement requirement) {
        Project project = getById(projectId);
        requirement.setProject(project);
        return requirementRepository.save(requirement);
    }

    @Transactional
    public ProjectRequirement updateRequirement(Long reqId, ProjectRequirement updated) {
        ProjectRequirement existing = requirementRepository.findById(reqId)
                .orElseThrow(() -> new NoSuchElementException("Requirement not found: " + reqId));
        existing.setLocation(updated.getLocation());
        existing.setNumberOfPositions(updated.getNumberOfPositions());
        existing.setRole(updated.getRole());
        return requirementRepository.save(existing);
    }

    @Transactional
    public void deleteRequirement(Long reqId) {
        if (!requirementRepository.existsById(reqId)) {
            throw new NoSuchElementException("Requirement not found: " + reqId);
        }
        requirementRepository.deleteById(reqId);
    }
}
